src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     * This should be useed if the dirty state of this instance can be controled by performing an operation rather than setting a new {@link DeferredProperty#value}.
+     *
+     * @param runnable The runnable which is run if we cleaned the object.
+     * @return <code>True</code> if the dirty flag changed to <code>true</code> from <code>false</code>.
+     */
+    public boolean clean(Runnable runnable) {
+        boolean cleaned = clean();
+        if (cleaned) {
+            runnable.run();
+        }
+        return cleaned;
+    }
+
+    /**
+     * Used to check if this instance is marked as dirty.
+     *
+     * @return True if this instance is dirty.
+     */
+    public boolean isDirty() {
+        return isDirty;
+    }
+
+    /**
+     * Used to check if this instance is _not_ marked as dirty.
+     *
+     * @return True if this instance is not dirty.
+     */
+    public boolean isClean() {
+        return !isDirty();
+    }
+
+    /**
+     * Set the {@link DeferredProperty#value} and cleans this instance by not marking it as dirty.
+     * Calling this function when this instance is not dirty will just set {@link DeferredProperty#value}.
+     *
+     * @param value The value to set this property {@link DeferredProperty#value} to.
+     * @return Returns <code>true</code> if the instance was cleaned by marking it as no longer dirty.
+     */
+    public boolean set(T value) {
+        this.value = value;
+        return clean();
+    }
+
+    /**
+     * Sets the {@link DeferredProperty#value} but only if this is dirty.
+     * Calling this function when this instance is not dirty will not set the value.
+     * However, the current value of this property will always be returned.
+     *
+     * @param value The value to set this property {@link DeferredProperty#value} to.
+     * @return Returns the value of this instance value even if no new assignment was made.
+     */
+    public T trySet(T value) {
+        if (isDirty()) {
+            set(value);
+        }
+        return get();
+    }
+
+    /**
+     * Sets the {@link DeferredProperty#value} but only if this is dirty.
+     * The new {@link DeferredProperty#value} would be the one returned by the <code>supplier</code>.
+     * Calling this function when this instance is not dirty will not set the value.
+     * However, the current value of this property will always be returned.
+     *
+     * @param supplier The supplier that returns the new value of {@link DeferredProperty#value}.
+     * @return Returns the value of this instance value even if no new assignment was made.
+     */
+    public T trySet(Supplier<T> supplier) {
+        return trySet(supplier.get());
+    }
+
+    /**
+     * If a supplier has be assigned and this is dirty then a new value is calcalted and returned.
+     * {@link Optional<T>} is used to highlight that if this instance is dirty then {@link Optional#empty()} is returned instead of <code>null</code>.
```

Comment:
```suggestion
     * {@link Optional<T>} is used to highlight that if this instance is dirty then {@link Optional#empty()} is returned instead of {@code null}.
```